### ISSUE TITLE
snapshot during training

### ISSUE THREAD

**Issue by Wulin-Tan** [danbider/lightning-pose#176]
Hi, LP team
in DLC training, some snapshots would be saved every few iterations. and we can make use of certain snapshots for prediction. 

Is there any function to do this in LP? 

let's say I train the model for 500 epochs. and finally I found that 300 epochs would be the best. Then I would like to pick 300 epochs model for later use, instead of training the model from beginning to 300 epochs.

in the LP, the prediction py file has a line:
```
model = load_model_from_checkpoint(cfg=cfg, ckpt_file=ckpt_file, eval=True)
```
that means I can load the specific ckpt file, I think if we can save the ckpt file every few steps like 50/100/200, that would be better.

**Comment by themattinthehatt** [comment]
I updated the code to allow you to do this. To enable this feature, add the following to your config file:
```
training:
   ckpt_every_n_epoch: 50
```
This will save a checkpoint every 50 epochs; update to whatever value you'd like, but note that `training.check_val_every_n_epoch` should evenly divide whichever number you choose. 

Please let me know if this works for you, and I'll close this issue.

**Comment by Wulin-Tan** [comment]
Hi, @themattinthehatt 
I tried it on LP 1.4.0 like:
```
  min_epochs: 30
  max_epochs: 30
  log_every_n_steps: 10
  check_val_every_n_epoch: 5
  gpu_id: 0
  rng_seed_data_pt: 0
  rng_seed_model_pt: 0
  lr_scheduler: multisteplr
  lr_scheduler_params:
    multisteplr:
      milestones:
      - 150
      - 200
      - 250
      gamma: 0.5
  ckpt_every_n_epochs: 10
```
but it did not work. only the final ckpt was saved. I listed all the ckpt files but only got one:
```
/root/autodl-tmp/DLC_LP/outputs/2024-06-18/12-43-09/tb_logs/DLC_LP/version_0/checkpoints/epoch=29-step=300.ckpt
```
any suggestion?

**Comment by themattinthehatt** [comment]
interesting, I checked this yesterday and it was working for me. How are you listing the checkpoints? Did you go into the actual checkpoints directory yourself to look? I'm not sure why that would've failed. I can test again on my side too.

**Comment by Wulin-Tan** [comment]
Hi, @themattinthehatt 
I run the code to list all the ckpt file:
```
import os

def find_ckpt_files(directory):
    ckpt_paths = []
    for root, _, files in os.walk(directory):
        for file in files:
            if file.endswith('.ckpt'):
                ckpt_paths.append(os.path.abspath(os.path.join(root, file)))
    return ckpt_paths

# Replace 'outputs' with your directory path
directory_path = 'outputs'
ckpt_paths = find_ckpt_files(directory_path)

# Print the absolute paths of all ckpt files
for path in ckpt_paths:
    print(path)

```

**Comment by Wulin-Tan** [comment]
Hi, @themattinthehatt 
I run the code to list all the ckpt file:
```
import os

def find_ckpt_files(directory):
    ckpt_paths = []
    for root, _, files in os.walk(directory):
        for file in files:
            if file.endswith('.ckpt'):
                ckpt_paths.append(os.path.abspath(os.path.join(root, file)))
    return ckpt_paths

# Replace 'outputs' with your directory path
directory_path = 'outputs'
ckpt_paths = find_ckpt_files(directory_path)

# Print the absolute paths of all ckpt files
for path in ckpt_paths:
    print(path)

```

**Comment by themattinthehatt** [comment]
@Wulin-Tan can you try again and watch the model directory during training? should be in `<model_dir>/tb_logs/<model_name>/version_0/checkpoints`. I was able to save the various checkpoints on my side using 1.4.0
![Screenshot from 2024-06-19 10-27-00](https://github.com/danbider/lightning-pose/assets/13826867/879771e2-2ee2-435b-a1e7-13ea074d4630)

**Comment by themattinthehatt** [comment]
Two questions:
1. can you send me the printout of your config file that is output to the terminal right before training starts? I want to make sure the checkpoint field is set properly in the config.
2. can you send me a screenshot of the function `get_callbacks()` in `lightning-pose/lightning_pose/utils/scripts.py`? Specifically I want to make sure you se the lines
```
if ckpt_model:
        if early_stopping:
            ckpt_callback = pl.callbacks.model_checkpoint.ModelCheckpoint(
                monitor="val_supervised_loss",
                mode="min",
            )
        else:
            # if ckpt_every_n_epochs is None, save after every validation step, but overwrite
            # if ckpt_every_n_epochs is not None, save separate checkpoint files
            ckpt_callback = pl.callbacks.model_checkpoint.ModelCheckpoint(
                monitor=None,
                every_n_epochs=ckpt_every_n_epochs,
                save_top_k=1 if ckpt_every_n_epochs is None else -1,
            )
```

**Comment by Wulin-Tan** [comment]
hi, @themattinthehatt 
here is the print out:
```
[2024-06-20 00:18:22,385][HYDRA] /root/miniconda3/lib/python3.8/site-packages/hydra/_internal/hydra.py:119: UserWarning: Future Hydra versions will no longer change working directory at job runtime by default.
See https://hydra.cc/docs/1.2/upgrades/1.1_to_1.2/changes_to_job_working_dir/ for more information.
  ret = run_job(

Our Hydra config file:
--------------------
data parameters
--------------------
image_orig_dims: {'height': 2160, 'width': 2160}
image_resize_dims: {'height': 256, 'width': 256}
data_dir: /root/autodl-tmp/DLC_LP
video_dir: /root/autodl-tmp/DLC_LP/videos
csv_file: CollectedData.csv
downsample_factor: 2
num_keypoints: 6
keypoint_names: ['snout', 'forepaw_L', 'forefaw_R', 'hindpaw_L', 'hindpaw_R', 'base']
mirrored_column_matches: None
columns_for_singleview_pca: [0, 1, 2, 3, 4, 5]

--------------------
training parameters
--------------------
imgaug: dlc
train_batch_size: 8
val_batch_size: 32
test_batch_size: 32
train_prob: 0.95
val_prob: 0.05
train_frames: 1
num_gpus: 1
num_workers: 16
early_stop_patience: 3
unfreezing_epoch: 20
min_epochs: 20
max_epochs: 20
log_every_n_steps: 10
check_val_every_n_epoch: 5
gpu_id: 0
rng_seed_data_pt: 0
rng_seed_model_pt: 0
lr_scheduler: multisteplr
lr_scheduler_params: {'multisteplr': {'milestones': [150, 200, 250], 'gamma': 0.5}}
ckpt_every_n_epochs: 5

--------------------
model parameters
--------------------
losses_to_use: ['pca_singleview', 'temporal']
backbone: resnet50_animal_ap10k
model_type: heatmap_mhcrnn
heatmap_loss_type: mse
model_name: DLC_LP

--------------------
dali parameters
--------------------
general: {'seed': 123456}
base: {'train': {'sequence_length': 32}, 'predict': {'sequence_length': 96}}
context: {'train': {'batch_size': 8}, 'predict': {'sequence_length': 48}}

--------------------
losses parameters
--------------------
pca_multiview: {'log_weight': 5.0, 'components_to_keep': 3, 'epsilon': None}
pca_singleview: {'log_weight': 5.0, 'components_to_keep': 0.99, 'epsilon': None}
temporal: {'log_weight': 5.0, 'epsilon': 20, 'prob_threshold': 0.05}

--------------------
callbacks parameters
--------------------
anneal_weight: {'attr_name': 'total_unsupervised_importance', 'init_val': 0.0, 'increase_factor': 0.01, 'final_val': 1.0, 'freeze_until_epoch': 0}
```

and the get_callbacks()
```
def get_callbacks(
    cfg: DictConfig,
    early_stopping=True,
    lr_monitor=True,
    ckpt_model=True,
    ckpt_every_n_epochs=None,
    backbone_unfreeze=True,
) -> List:

    callbacks = []

    if early_stopping:
        early_stopping = pl.callbacks.EarlyStopping(
            monitor="val_supervised_loss",
            patience=cfg.training.early_stop_patience,
            mode="min",
        )
        callbacks.append(early_stopping)
    if lr_monitor:
        lr_monitor = pl.callbacks.LearningRateMonitor(logging_interval="epoch")
        callbacks.append(lr_monitor)
    if ckpt_model:
        if early_stopping:
            ckpt_callback = pl.callbacks.model_checkpoint.ModelCheckpoint(
                monitor="val_supervised_loss",
                mode="min",
            )
        else:
            # if ckpt_every_n_epochs is None, save after every validation step, but overwrite
            # if ckpt_every_n_epochs is not None, save separate checkpoint files
            ckpt_callback = pl.callbacks.model_checkpoint.ModelCheckpoint(
                monitor=None,
                every_n_epochs=ckpt_every_n_epochs,
                save_top_k=1 if ckpt_every_n_epochs is None else -1,
            )
        callbacks.append(ckpt_callback)
    if backbone_unfreeze:
        transfer_unfreeze_callback = pl.callbacks.BackboneFinetuning(
            unfreeze_backbone_at_epoch=cfg.training.unfreezing_epoch,
            lambda_func=lambda epoch: 1.5,
            backbone_initial_ratio_lr=0.1,
            should_align=True,
            train_bn=True,
        )
        callbacks.append(transfer_unfreeze_callback)
    # we just need this callback for unsupervised models
    if (cfg.model.losses_to_use != []) and (cfg.model.losses_to_use is not None):
        anneal_weight_callback = AnnealWeight(**cfg.callbacks.anneal_weight)
        callbacks.append(anneal_weight_callback)

    return callbacks
```

**Comment by themattinthehatt** [comment]
can you share where `get_callbacks()` is called in train.py? should look like this (L73):
```
    callbacks = get_callbacks(
        cfg,
        early_stopping=False,
        ckpt_every_n_epochs=cfg.training.get("ckpt_every_n_epochs", None)
    )
```

**Comment by Wulin-Tan** [comment]
hi, @themattinthehatt 
my ```get_callbacks()``` is like:
```
    # early stopping, learning rate monitoring, model checkpointing, backbone unfreezing
    callbacks = get_callbacks(cfg, early_stopping=False)
```
and it does not include ckpt_every_n_epochs. So I follow your code:
```
    callbacks = get_callbacks(
        cfg,
        early_stopping=False,
        ckpt_every_n_epochs=cfg.training.get("ckpt_every_n_epochs", None)
    )
```
and it works!

I also tried your new version of ```train_hydra.py``` like:
```
"""Example model training script."""

import hydra
from omegaconf import DictConfig

from lightning_pose.train import train


@hydra.main(config_path="configs", config_name="config_mirror-mouse-example")
def train_model(cfg: DictConfig):
    """Main fitting function, accessed from command line.

    To train a model on the example dataset provided with the Lightning Pose package with this
    script, run the following command from inside the lightning-pose directory
    (make sure you have activated your conda environment):

    ```
    python scripts/train_hydra.py
    ```

    Note there are no arguments - this tells the script to default to the example data.


    To train a model on your own dataset, overwrite the default config_path and config_name args:

    ```
    python scripts/train_hydra.py --config-path=<PATH/TO/YOUR/CONFIGS/DIR> --config-name=<CONFIG_NAME.yaml>  # noqa
    ```

    For more information on training models, see the docs at
    https://lightning-pose.readthedocs.io/en/latest/source/user_guide/training.html

    """

    train(cfg)


if __name__ == "__main__":
    train_model()
```
It also works.

by the way, why do you set early_stopping false? I think it is useful and in the config file you also have ```early_stop_patience```

**Comment by themattinthehatt** [comment]
@Wulin-Tan you should have those updates with the latest code - did you run `git pull` recently?

We have used early stopping in the past but actually the way it is implemented right now has some issues - the validation data (which is used for early stopping) uses the same image augmentation pipeline as the training data, which means that the validation data is actually different every time. So a low validation loss might mean that the model is doing better, or that batch just so happened to have some easier augmentations (by chance). This is on our list of things to fix, but for the time being we set the default to no early stopping. I should note that the early stopping still performs reasonably this way, but it's not perfect.

**Comment by hummuscience** [comment]
@themattinthehatt with the recent correction of the augmentation for the validation, is it safe to say that one should use early stopping by default? I am worried about over-trained models. 

For example, a convergence at around 300 epochs with a "best" ckpt saved there and then a slight improvement at a single validation 300 epochs later. That is a bit worrisome, no?

**Comment by themattinthehatt** [comment]
You could turn on early stopping and it would definitely prevent that slight improvement 300 epochs later, but it's not clear to me if that model would be more overfit than the model checkpointed at epoch 300. Typically I set the max number of epochs to 300 to prevent the model from training too long anyways. You could try checkpointing a model at 300 and 600 epochs and see if their predictions are noticeably different.
